netfx_loader/DomainData.cs

```diff
@@ -1,54 +1,94 @@
+            var assembly = domain.Load(AssemblyName.GetAssemblyName(assemblyPath));
+            var type = assembly.GetType(typeName, throwOnError: true);
+            var deleg = Delegate.CreateDelegate(typeof(EntryPoint), type, function);
+            IntPtr result = Marshal.GetFunctionPointerForDelegate(deleg);
+            return result;
+        }
+    }
 
+    class DomainData : IDisposable
+    {
         bool _disposed = false;
 
         public AppDomain Domain { get; }
-        public Dictionary<(string, string, string), EntryPoint> _delegates;
+        public Dictionary<(string, string, string), IntPtr> _functors;
+        public HashSet<string> _resolvedAssemblies;
 
         public DomainData(AppDomain domain)
         {
             Domain = domain;
-            _delegates = new Dictionary<(string, string, string), EntryPoint>();
+            _functors = new Dictionary<(string, string, string), IntPtr>();
+            _resolvedAssemblies = new HashSet<string>();
         }
 
-        public EntryPoint GetEntryPoint(string assemblyPath, string typeName, string function)
+        private void installResolver(string assemblyPath)
+        {
+            var assemblyName = AssemblyName.GetAssemblyName(assemblyPath).Name;
+            if (_resolvedAssemblies.Contains(assemblyName))
+                return;
+            _resolvedAssemblies.Add(assemblyName);
+
+            AppDomain.CurrentDomain.AssemblyResolve += (sender, args) =>
+            {
+                if (args.Name.Contains(assemblyName))
+                    return Assembly.LoadFrom(assemblyPath);
+                return null;
+            };
+        }
+
+        public IntPtr GetFunctor(string assemblyPath, string typeName, string function)
         {
             if (_disposed)
                 throw new InvalidOperationException("Domain is already disposed");
 
-            var key = (assemblyPath, typeName, function);
+            installResolver(assemblyPath);
 
-            EntryPoint result;
+            var key = (assemblyPath, typeName, function);
 
-            if (!_delegates.TryGetValue(key, out result))
+            IntPtr result;
+            if (!_functors.TryGetValue(key, out result))
             {
-                var assembly = Domain.Load(AssemblyName.GetAssemblyName(assemblyPath));
-                var type = assembly.GetType(typeName, throwOnError: true);
+                Domain.SetData("_assemblyPath", assemblyPath);
+                Domain.SetData("_typeName", typeName);
+                Domain.SetData("_function", function);
 
-                Print($"Loaded type {type}");
-                result = (EntryPoint)Delegate.CreateDelegate(typeof(EntryPoint), type, function);
-
-                _delegates[key] = result;
+                Domain.DoCallBack(new CrossAppDomainDelegate(DomainSetup.StoreFunctorFromDomainData));
+                result = (IntPtr)Domain.GetData("_thisFunctor");
```

Comment:
1. It might make sense to unset this after it is read and check the value here
2. This is not at all thread-safe. The simplest approach is probably to lock here in C#.

Reply:
We can't unset the functor since we don't know when the caller of `pyclr_get_function` will actually invoke the functor.
It is only when the domain is closed that we can do this, and by closing the domain all the domain data will be reclaimed anyways.

But we can certainly unset the inputs.